doc/man1/roarvorbis.1
.\" roarvorbis.1:

.TH "roarvorbis" "1" "January 2010" "RoarAudio" "System User's Manual: roarvorbis"

.SH NAME

roarvorbis \- Ogg Vorbis player for RoarAudio

.SH SYNOPSIS

roarvorbis [OPTIONS]... FILE

.SH DESCRIPTION

Plays an Ogg Vorbis file showing it's metadata.

.SH "OPTIONS"

.TP
\fB--server SERVER\fR
Set server hostname

.TP
\fB--help\fR
Show this help

.TP
\fB--vclt-out FILE\fR
Writes VCLT file

.SH "SEE ALSO"
\fBroartips\fR(7),
\fBlibroar\fR(7),
\fBRoarAudio\fR(7).

.SH "HISTORY"

For history information see \fBRoarAudio\fR(7).

.\" ll


ChangeLog
v. 0.3beta3 - ?
	* added some aRts/NAS/YIFF/PulseAudio tools to compbins
	* added support for comp libs to roar-config
	* added support for uni-directional streams in roarinterconnect
	* write much better support for config, including codec config
	* reduced the TCP overhead produced by speex
	* added support for speex preprocessor
	* get names of complibs from table, do not guess them
	* a lot small fixes
	* moved most of the mixer stuff out of roard into libroardsp
	* added flags immutable and enhance
	* test for signal.h
	* Corrected (DSTR) HTTP Protocol support so servers without headerflush
	  should now work, too.
	* added protocol support for icy
	* Added version numbers to lib names
	* updated manpages
	* wrote OSS emulation

v. 0.3beta2 - Mon Oct 05 2009 19:48 CEST
	* Added prethru flag
	* Added support to read info about client's network node
	* Ported libroaresd, libroarartsc
	* Fixed some problems with EsounD emulation
	* added support for compatibility binarys
	* added poll() support to libroarsndio
	* reject streams of unsupported codec

v. 0.3beta1 - Sat Sep 26 2009 25:50 CEST
	* incremented nummber of streams per client fron 4 to 8
	* RIFF WAVE codecfilter uses virtual streams
	* another set of changed names of parameters in roarctl
	* Interface to read byteorder of a client from roard
	* implemented a 'state of stream' var
	* cleanup, small fixes
	* added VIO interface to transcoder interface
	* ask codecfilter of parent in case of new child stream
	* added 'subsystem' complex: multi subsystem data on single streams
	* added basic /Radio Data and Trasmitter Control System/.
	* better support for stereo upmixing
	* make audio group a optional thing
	* better support for win32
	* wrote WMM driver
	* better multi protocol support
	* Support for server side EsounD emulation on win32

v. 0.3beta0 - Sun Sep 06 2009 22:42 CEST
	* added new stream directions for MIDI and light control
	* cleanup of client stream struct
	* added basic light control subsystem
	* added DMX512 driver
	* wrote basic MIDI subsystem
	* added flag 'mute'
	* added flag 'mmap'
	* added lib libroarmidi and libroarlight
	* added OpenSLP support
	* rewote fillmixbuer2 code
	* added a roar_conv2() with a more clear API
	* added support to write pidfile
	* added support for encoding 32 bit Vorbis streams
	* added transcoding features in libroardsp
	* added tool 'roarphone'
	* added tool 'roarinterconnect'
	* added simple raw data subsystem
	* added working support for THRU streams
	* changed names of parameters in roarctl
	* added lib libroareio - Extended IO for RoarAudio
	* Update order of default drivers, now it is: oss, ao, esd, null
	* cleanup of old sysio driver interface
	* added server side EsounD emulation
	* added multi homed support to server
	* fixed a long outstanding bug with pa_path_get_filename()

v. 0.2 - Thu May 21 2009 21:00
	* do a some more testings for functions not existing on win32
	* init winsock
	* done a lot fixes in order to get libroar and some clients work on win32
	* done a lot cleanup on configure script

v. 0.2rc2 - Sun May 17 2009 25:44
	* corrected check for GNU peercred
	* added basic logging support to configure
	* fixed plugin configures
	* added and fixed semi-install target on plugins
	* call plugin configures from main configure
	* include plugins in Makefile known to work from configure
	* fixed Makefiles of plugins: added -I../../include, -L../../lib
	* Make use of pkg-config in plugin configures
	* corrected possition of -ls on linker commands in most Makefiles
	* added more explicide linker commands so most things now build
	  with classic BSD make.
	* devided into components: libs, roard, clients, doc
	* add toption to set roard's cycle freq
	* moved a lot things in plugin's configure scripts to a commone one

v. 0.2rc1 - Wed May 13 2009 12:05
	* Started to port to OpenSolaris
	* fixed bug in default based sockets, make HTTP VIO from non default port work
	* got libroarsndio build under OpenBSD 4.5
	* do not depend tests in configure on roaraudio.h: fixes some checks on *BSD
	* renamed member 'fh' of connection struct to '__fh', this breaks
	  all code depending on it so we see how it behaves in case
	  we replace this with a real VIO object in near future
	* started VIO 2popen interface
	* fixed exec() on systems with sizeof(int) != sizeof(char*)
	* do not do unnessesery calls to fdatasync() on sockets
	* link roarvorbis against libogg and libvorbis
	* remove need of GNU/Tar like -z on remote system from send-roaraudio
	* added support for hwmixer flag on OSS and sndio driver
	* fiexed some typos

v. 0.2beta3 - Fri May 01 2009 15:15
	* added esdfilt options to roarfilt, esdplay options to roarcatplay
	* added short options to roarcat/mon/filt as stated in the manpage
	* updated manpages
	* also build libroar{,dsp}.a
	* Changed VIO CTLs in a binary inkompatible way
	* added support for autoconf flag to OSS driver
	* added driver for OpenBSD sndio interface
	* started libroarsndio
	* added code for plugins: audacirus, mplayer, gst10
	* fixed some building problems
	* added roarify and s start for it's manpage
	* added stream flags pause and hwmixer

v. 0.2beta2 - Fri Mar 27 2009 21:20
	* added (dummy) flags autoconf and cleanmeta
	* added streams_ctl() to server
	* added support for primary flag as -oO, support for blocks parameter
	* added (GnuPG) pinentry support.
	* added support for Ogg Vorbis encoder to set meta data :)
	* fixed some (minor) bugs, including segfault in vio
	* added a lot things to the vio interface
	* added gzip and simple OpenPGP support to vio interface
	* added a simple pointer stack.
	* added (OpenSSH) askpass support.
	* added vio pipe support, currectly pipe() and socketpair() based onse.
	* test for OpenSSL
	* added support to VIO open OpenSSL BIOs
	* added support to read a string and convert it into a VIO stack
	* added simple VIO stack as VIO pass object
	* started cleaner VIO Socket implementation

roarclients/roarvorbis.c
//roarvorbis.c:

/*
 *      Copyright (C) Philipp 'ph3-der-loewe' Schafft - 2008
 *
 *  This file is part of roarclients a part of RoarAudio,
 *  a cross-platform sound system for both, home and professional use.
 *  See README for details.
 *
 *  This file is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License version 3
 *  as published by the Free Software Foundation.
 *
 *  RoarAudio is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this software; see the file COPYING.  If not, write to
 *  the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */

#include <roaraudio.h>

#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#ifdef ROAR_HAVE_LIBVORBISFILE
#include <vorbis/codec.h>
#include <vorbis/vorbisfile.h>
#endif

#ifdef _WIN32
#include <io.h>
#include <fcntl.h>
#endif

#define BUFSIZE 1024

void usage (void) {
 printf("roarvorbis [OPTIONS]... FILE\n");

 printf("\nOptions:\n\n");

 printf("  --server SERVER    - Set server hostname\n"
        "  --help             - Show this help\n"
        "  --vclt-out FILE    - Writes VCLT file\n"
       );

}


#ifdef ROAR_HAVE_LIBVORBISFILE
FILE * open_http (char * file) {
#ifdef ROAR_HAVE_BIN_WGET
 char cmd[1024];

 snprintf(cmd, 1023, ROAR_HAVE_BIN_WGET " -qO - '%s'", file);

 return popen(cmd, "r");
#else
 return NULL;
#endif
}

int update_stream (struct roar_connection * con, struct roar_stream * s, int * out, OggVorbis_File * vf, char * file, struct roar_audio_info * info, struct roar_vio_calls * vclt) {
 vorbis_info *vi = ov_info(vf, -1);
 int    bits     = 16;
 int    codec    = ROAR_CODEC_DEFAULT;
 char **ptr = ov_comment(vf, -1)->user_comments;
 char key[ROAR_META_MAX_NAMELEN], value[LIBROAR_BUFFER_MSGDATA] = {0};
 int j, h = 0;
 struct roar_meta   meta;
 int need_new_stream = 0;
 int meta_ok;

 fprintf(stderr, "\n");

 if ( vclt != NULL ) {
  roar_vio_printf(vclt, "AUDIOINFO=rate:%iHz, channels:%i\n", vi->rate, vi->channels);
 }

 if ( *out == -1 ) {
  need_new_stream = 1;
 } else if ( info->rate != vi->rate || info->channels != vi->channels ) {
  need_new_stream = 1;
 }

 if ( need_new_stream ) {
  if ( *out != -1 )
  close(*out);

  fprintf(stderr, "Audio: %i channel, %liHz\n\n", vi->channels, vi->rate);

  info->rate     = vi->rate;
  info->channels = vi->channels;

  if ( (*out = roar_simple_new_stream_obj(con, s, vi->rate, vi->channels, bits, codec, ROAR_DIR_PLAY)) == -1 ) {
   roar_disconnect(con);
   return -1;
  }
 }


 meta.value = value;
 meta.key[0] = 0;
 meta.type = ROAR_META_TYPE_NONE;

 roar_stream_meta_set(con, s, ROAR_META_MODE_CLEAR, &meta);

 if ( strncmp(file, "http:", 5) == 0 )
  meta.type = ROAR_META_TYPE_FILEURL;
 else
  meta.type = ROAR_META_TYPE_FILENAME;


 strncpy(value, file, LIBROAR_BUFFER_MSGDATA-1);
 value[LIBROAR_BUFFER_MSGDATA-1] = 0;
 roar_stream_meta_set(con, s, ROAR_META_MODE_SET, &meta);

 while(*ptr){
  meta_ok = 1;

   for (j = 0; (*ptr)[j] != 0 && (*ptr)[j] != '='; j++) {
    if ( j == ROAR_META_MAX_NAMELEN ) {
     ROAR_ERR("update_stream(*): invalid meta data: meta data key too long");
     meta_ok = 0;
     j = 0;
     break;
    }
    key[j] = (*ptr)[j];
   }
   key[j] = 0;

   if ( meta_ok ) {
    for (j++, h = 0; (*ptr)[j] != 0 && (*ptr)[j] != '='; j++) {
     if ( h == LIBROAR_BUFFER_MSGDATA ) {
      ROAR_ERR("update_stream(*): invalid meta data: meta data value for key '%s' too long", key);
      meta_ok = 0;
      h = 0;
      break;
     }
     value[h++] = (*ptr)[j];
    }
    value[h]   = 0;
   }

   if ( meta_ok ) {
    fprintf(stderr, "Meta %-16s: %s\n", key, value);

    if ( vclt != NULL ) {
     roar_vio_printf(vclt, "%s=%s\n", key, value);
    }

    meta.type = roar_meta_inttype(key);
    if ( meta.type != -1 )
     roar_stream_meta_set(con, s, ROAR_META_MODE_SET, &meta);
   }

   ptr++;
 }

 *value      = 0;
 meta.key[0] = 0;
 meta.type   = ROAR_META_TYPE_NONE;
 roar_stream_meta_set(con, s, ROAR_META_MODE_FINALIZE, &meta);

 if ( vclt != NULL ) {
  roar_vio_printf(vclt, "==\n");
 }

 return 0;
}

#endif

int main (int argc, char * argv[]) {
 struct roar_vio_calls vclt;
 struct roar_vio_defaults def;
#ifndef ROAR_HAVE_LIBVORBISFILE
 fprintf(stderr, "Error: no Vorbis support!\n");
 return 1;
#else
 char * server   = NULL;
 char * file     = NULL;
 char * vcltfile = NULL;
 char * k;
 int    i;
 FILE * in;
 int    out = -1;
 struct roar_connection con;
 struct roar_stream     s;
 OggVorbis_File vf;
 int eof=0;
 int current_section = -1;
 int last_section = -1;
 struct roar_audio_info info;
 char pcmout[4096];


 for (i = 1; i < argc; i++) {
  k = argv[i];

  if ( strcmp(k, "--server") == 0 ) {
   server = argv[++i];
  } else if ( strcmp(k, "--vclt-out") == 0 ) {
   vcltfile = argv[++i];
  } else if ( strcmp(k, "--help") == 0 ) {
   usage();
   return 0;
  } else if ( file == NULL ) {
   file = k;
  } else {
   fprintf(stderr, "Error: unknown argument: %s\n", k);
   usage();
   return 1;
  }
 }

 roar_libroar_set_server(server);

 if ( file == NULL ) {
  ROAR_ERR("No filename given.");
  return 1;
 }

 if ( roar_simple_connect(&con, server, "roarvorbis") == -1 ) {
  ROAR_DBG("roar_simple_play(*): roar_simple_connect() faild!");
  return 1;
 }

 if ( strncmp(file, "http:", 5) == 0 ) {
  in = open_http(file);
 } else {
  in = fopen(file, "rb");
 }

 if ( in == NULL ) {
  roar_disconnect(&con);
  return 1;
 }

#ifdef _WIN32
  _setmode(_fileno(in), _O_BINARY);
#endif

 if(ov_open(in, &vf, NULL, 0) < 0) {
  fprintf(stderr,"Input does not appear to be an Ogg bitstream.\n");
  roar_disconnect(&con);
  return 1;
 }

 if ( vcltfile != NULL ) {
  if ( roar_vio_dstr_init_defaults(&def, ROAR_VIO_DEF_TYPE_NONE, O_WRONLY|O_CREAT|O_APPEND, 0644) == -1 )
   return 1;
  if ( roar_vio_open_dstr(&vclt, vcltfile, &def, 1) == -1 ) {
   fprintf(stderr, "Error: can not open file: %s: %s\n", k, strerror(errno));
   return 1;
  }
 }

// if ( update_stream(&con, &s, &out, &vf, file) == -1 )
//  return 1;

 while (!eof) {
  long ret = ov_read(&vf, pcmout, sizeof(pcmout), 0, 2, 1, &current_section);

  if ( last_section != current_section )
   if ( update_stream(&con, &s, &out, &vf, file, &info, vcltfile == NULL ? NULL : &vclt) == -1 )
    return 1;

  last_section = current_section;

  if (ret == 0) {
   /* EOF */
   eof=1;
  } else if (ret < 0) {
   /* error in the stream.  Not a problem, just reporting it in
      case we (the app) cares.  In this case, we don't. */
  } else {
     /* we don't bother dealing with sample rate changes, etc, but
        you'll have to */
//    write(out, pcmout, ret);
   roar_stream_send_data(&con, &s, pcmout, ret);
  }
 }

  ov_clear(&vf);

// fclose(in);
 close(out);
 roar_disconnect(&con);

 if ( vcltfile != NULL ) {
  roar_vio_close(&vclt);
 }

 return 0;
#endif
}

//ll
